### PR TITLE
[MAID-3161] feat/seed: add ability to generate sign keypair using a given seed 

### PR DESCRIPTION
**Pull request by mrcnski**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -155,10 +161,30 @@ pub fn init_with_rng<T: Rng>(rng: &mut T) -> Result<(), Error> {
...
 /// Produces a 256-bit crypto hash out of the provided `data`.
-pub fn hash(data: &[u8]) -> [u8; 32] {
+pub fn hash(data: &[u8]) -> [u8; HASH_BYTES] {
     hashing_impl::sha3_256(data)
 }
 
+/// Seed structure used to generate sign and encrypt keypairs deterministically.
+#[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+pub struct Seed {
+    seed: sign::Seed,
```
Comment (ustulation) [inline src/lib.rs:217]: `seed.seed` feels strange. Calling it `inner` is better ? Also check [review comment here](https://github.com/maidsafe/safe_crypto/pull/26#discussion_r219439233) pls.

File: src/lib.rs [hunk 2]
```diff
@@ -734,34 +827,71 @@ mod tests {
...
+        let (pk2, sk2) = gen_sign_keypair_from_seed(&seed);
+
+        assert_eq!(pk1, pk2);
+        assert_eq!(sk1, sk2);
+
+        let sig1 = sk1.sign_detached(&data1);
+        let sig2 = sk2.sign_detached(&data2);
+
+        assert!(pk1.verify_detached(&sig1, &data1));
+        assert!(pk2.verify_detached(&sig2, &data2));
```
Comment (ustulation) [inline src/lib.rs:923]: We normally also `assert` a failure for the "misuse" case for comprehensiveness.

File: src/mock_crypto.rs [hunk 3]
```diff
@@ -234,6 +236,18 @@ pub(crate) mod crypto_impl {
...
+            /// Mock seed.
+            #[derive(Debug, PartialEq, Eq, Serialize, Deserialize, Clone)]
+            pub(crate) struct Seed(pub(crate) [u8; SEEDBYTES]);
+
+            /// Generate mock public and corresponding secret key using a seed.
+            pub(crate) fn keypair_from_seed(seed: &Seed) -> (PublicKey, SecretKey) {
+                let pub_key = seed.0;
+                let mut sec_key = [0u8; 64];
+                sec_key[0..32].clone_from_slice(&seed.0); // simply get 64 byte array
+                (PublicKey(pub_key), SecretKey(sec_key))
```
Comment (ustulation) [inline src/mock_crypto.rs:266]: Don't we normally make sec-key == pub-key in mock ? pls go over the rest of mock file and confirm.

Also don't see any test cases for this.

**Review by ustulation — CHANGES_REQUESTED** [review]

File: src/mock_crypto.rs (hunk 3 above)
Comment (mrcnski) [reply src/mock_crypto.rs:266]: Well the secret key is 64 bytes while the public key is 32, but the first half of the secret key is equal to the public key, yeah. Added test cases.

File: src/lib.rs (hunk 1 above)
Comment (mrcnski) [reply src/lib.rs:217]: Replied to this on the linked review comment. We'll leave the names as they are.

File: src/lib.rs (hunk 2 above)
Comment (mrcnski) [reply src/lib.rs:923]: Added!

**Review by ustulation — APPROVED** [review]

**Pull request merged**